File: src/exercise_service_py/exercise_service_py/waypoint_manager_server.py
import rclpy
from rclpy.node import Node
from base_interfaces.srv import ManageWaypoint

class WaypointManagerServer(Node):
    def __init__(self):
        super().__init__("waypoint_manager_server_py")
        self.get_logger().info("服务端创建成功!(Python)")
        self.server_ = self.create_service(ManageWaypoint, "manage_waypoint", self.waypoint_manager_callback)
        self.waypoint_ = []

    def waypoint_manager_callback(self, req, res):
        operation = req.operation
        waypoint_name = req.waypoint_name
        x = req.x
        y = req.y
        yaw = req.yaw

        if operation == "add":
            if not waypoint_name:
                self.set_failure_response(res, "航点名称不能为空!")
            elif self.has_waypoint(waypoint_name):
                self.set_failure_response(res, "存在同名航点!")
            elif len(self.waypoint_) >= 5:
                self.set_failure_response(res, "当前航点数量已达到 5 个!")
            elif (
                x < -10.0 - 1e-6 or x > 10.0 + 1e-6 or
                y < -10.0 - 1e-6 or y > 10.0 + 1e-6 or
                yaw < -3.14 - 1e-6 or yaw > 3.14 + 1e-6
            ):
                self.set_failure_response(res, "航点参数不合法!")
            else:
                new_waypoint = {"waypoint_name": waypoint_name, "x": x, "y": y, "yaw": yaw}
                self.waypoint_.append(new_waypoint)
                res.success = True
                res.message = "航点添加成功!"
                self.get_logger().info("航点添加成功!")
        elif operation == "delete":
            index = self.find_waypoint_index(waypoint_name)
            if index == -1:
                self.set_failure_response(res, "该航点不存在!")
            else:
                self.waypoint_.pop(index)
                res.success = True
                res.message = "航点删除成功!"
                self.get_logger().info("航点删除成功!")
        elif operation == "query":
            res.success = True
            res.message = "航点查询成功!"
            self.get_logger().info("航点查询成功!")
        elif operation == "clear":
            self.waypoint_.clear()
            res.success = True
            res.message = "航点清空成功!"
            self.get_logger().info("航点清空成功!")
        else:
            self.set_failure_response(res, "未知航点!")

        self.fill_waypoints_response(res)
        return res

    def set_failure_response(self, res, message):
        res.success = False
        res.message = message
        self.get_logger().warn(f"{message}")

    def has_waypoint(self, waypoint_name):
        for waypoint in self.waypoint_:
            if waypoint_name == waypoint["waypoint_name"]:
                return True
        return False
    
    def find_waypoint_index(self, waypoint_name):
        for i in range(len(self.waypoint_)):
            if waypoint_name == self.waypoint_[i]["waypoint_name"]:
                return i
        return -1

    def fill_waypoints_response(self, res):
        res.waypoint_names = []
        res.xs = []
        res.ys = []
        res.yaws = []

        for waypoint in self.waypoint_:
            res.waypoint_names.append(waypoint["waypoint_name"])
            res.xs.append(waypoint["x"])
            res.ys.append(waypoint["y"])
            res.yaws.append(waypoint["yaw"])
        
        res.total_count = len(self.waypoint_)

def main():
    rclpy.init()
    server = WaypointManagerServer()

    try:
        rclpy.spin(server)
    except KeyboardInterrupt:
        pass
    finally:
        server.destroy_node()

        if rclpy.ok():
            rclpy.shutdown()

if __name__ == "__main__":
    main()

File: src/exercise_service_cpp/src/waypoint_manager_server.cpp
#include "base_interfaces/srv/manage_waypoint.hpp"
#include "rclcpp/rclcpp.hpp"
#include <cmath>
#include <string>
#include <vector>

using base_interfaces::srv::ManageWaypoint;
using std::placeholders::_1;
using std::placeholders::_2;

struct Waypoint {
    std::string waypoint_name;
    float x;
    float y;
    float yaw;
};

class WaypointManagerServer : public rclcpp::Node {
public:
    WaypointManagerServer() : Node("waypoint_manager_server_cpp") {
        RCLCPP_INFO(this->get_logger(), "服务端创建成功!");
        server_ = this->create_service<ManageWaypoint>("manage_waypoint", std::bind(&WaypointManagerServer::waypoint_manager_callback, this, _1, _2));
    }

private:
    void waypoint_manager_callback(const ManageWaypoint::Request::SharedPtr request, ManageWaypoint::Response::SharedPtr response) {
        const std::string &operation = request->operation;
        const std::string &waypoint_name = request->waypoint_name;
        const float x = request->x;
        const float y = request->y;
        const float yaw = request->yaw;

        if (operation == "add") {
            if (waypoint_name.empty()) {
                set_failure_response(response, "航点名称不能为空!");
            } else if (has_waypoint(waypoint_name)) {
                set_failure_response(response, "存在同名航点!");
            } else if (waypoints_.size() >= 5) {
                set_failure_response(response, "当前航点数量已达到 5 个!");
            } else if (!std::isfinite(x) || !std::isfinite(y) || !std::isfinite(yaw) || x < -10.0F || x > 10.0F || y < -10.0F || y > 10.0F || yaw < -3.14F || yaw > 3.14F) {
                set_failure_response(response, "航点参数不合法!");
            } else {
                waypoints_.push_back({waypoint_name, x, y, yaw});
                response->success = true;
                response->message = "航点添加成功!";
                RCLCPP_INFO(this->get_logger(), "航点添加成功!");
            }
        } else if (operation == "delete") {
            const int index = find_waypoint_index(waypoint_name);
            if (index == -1) {
                set_failure_response(response, "该航点不存在!");
            } else {
                waypoints_.erase(waypoints_.begin() + index);
                response->success = true;
                response->message = "航点删除成功!";
                RCLCPP_INFO(this->get_logger(), "航点删除成功!");
            }
        } else if (operation == "query") {
            response->success = true;
            response->message = "航点查询成功!";
            RCLCPP_INFO(this->get_logger(), "航点查询成功!");
        } else if (operation == "clear") {
            waypoints_.clear();
            response->success = true;
            response->message = "航点清空成功!";
            RCLCPP_INFO(this->get_logger(), "航点清空成功!");
        } else {
            set_failure_response(response, "未知操作!");
        }

        fill_waypoints_response(response);
    }

    bool has_waypoint(const std::string &waypoint_name) const {
        for (const auto &waypoint : waypoints_) {
            if (waypoint.waypoint_name == waypoint_name) {
                return true;
            }
        }

        return false;
    }

    int find_waypoint_index(const std::string &waypoint_name) const {
        for (size_t i = 0; i < waypoints_.size(); ++i) {
            if (waypoints_[i].waypoint_name == waypoint_name) {
                return static_cast<int>(i);
            }
        }

        return -1;
    }

    void set_failure_response(const ManageWaypoint::Response::SharedPtr response, const std::string &message) {
        response->success = false;
        response->message = message;
        RCLCPP_WARN(this->get_logger(), "%s", message.c_str());
    }

    void fill_waypoints_response(const ManageWaypoint::Response::SharedPtr response) {
        response->waypoint_names.clear();
        response->xs.clear();
        response->ys.clear();
        response->yaws.clear();

        for (const auto &waypoint : waypoints_) {
            response->waypoint_names.push_back(waypoint.waypoint_name);
            response->xs.push_back(waypoint.x);
            response->ys.push_back(waypoint.y);
            response->yaws.push_back(waypoint.yaw);
        }

        response->total_count = static_cast<int32_t>(waypoints_.size());
    }

    rclcpp::Service<ManageWaypoint>::SharedPtr server_;
    std::vector<Waypoint> waypoints_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<WaypointManagerServer>());
    rclcpp::shutdown();

    return 0;
}
